Update software.md with web ui docs (#99)
ROAR-QUTRC/perseus-v2: docs/source/architecture/software.md

This library is not used for ROS2 communications, it exists for scenarios such as a creating a ROS2 driver node which needs to interface with a specific sensor via ethernet.
:::

## Perseus-UI

The Perseus-UI is a web UI that is almost entirely independent of the rover core software. The UI is a SvelteKit + Vite app with an injected Node.js server. There are two main components to Perseus-UI:

- Widgets -> Typically a single svelte component that provide an interface for a specific functionality on the rover.
- Layouts -> A group of widgets that are regularly used together.

### Running Perseus-UI

The Rover UI uses SvelteKit meaning that it does not need a server to be run and can be started by running these commands in the root directory:

> ```bash

Reviewer: ```suggestion
```shell
```
If it's interactive commands to run, `shell` is the preferred "language" instead of `bash`, which is intended to document, well, bash scripts.

yarn # This installs dependencies
yarn build # Build the server
yarn start # Runs the built server on port 3000
```

**Note:** Any hydration related errors can be ignored

:::{note}
If you don't have node/yarn I'd recommend these steps:

> Instal node version manager with this script (you will need to restart your shell once its installed):

Reviewer: ```suggestion
Install node version manager with this script (you will need to restart your shell once its installed):
```

```
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash
```

Install the latest lts version of node:

```
nvm install --lts
```

If you have other versions of node already installed check the [nvm docs](https://github.com/nvm-sh/nvm) on how to setup environments or change the active node version.

Finally, install yarn with:

```
npm install --global yarn

> ```

Reviewer: I'll let this stand for now because I don't have time to fix it, but this should be managed with Nix, _not_ with an external tool (npm in this case). IE, `yarn` gets added to a `.#web-ui` shell or something similar. If people set this up then use Nix afterward it _should_ be fine, since Nix should override the environment variables/PATH, but it might end up being an issue. Will need testing.

:::

### Developing for the UI

#### Client

> The UI library used is [shadcn-svelte](https://next.shadcn-svelte.com), for ROS connections use [roslibjs](https://robotwebtools.github.io/roslibjs/index.html).

Reviewer: ```suggestion
The UI library used is [shadcn-svelte](https://next.shadcn-svelte.com), with [roslibjs](https://robotwebtools.github.io/roslibjs/index.html) used for communication with the ROS stack.
```

#### Server

> SvelteKit provides a number of methods of assisting with server communication however I would recommend using the node server and websockets to prevent weird bugs from components rerendering. Since Svelte doesn't provide a native interface for creating custom node servers an injected server is used. For development and testing use write your node server code in the vite plugin inside `vite.config.ts` and copy this code into `web-ui/src/server/server.js` for production (alternatively create a new file that exports a function and call it in both then develop from that file).

Reviewer: ```suggestion
SvelteKit provides a number of methods of assisting with server communication, however I would recommend using the node server and WebSockets to prevent weird bugs from components re-rendering. Since Svelte doesn't provide a native interface for creating custom node servers an injected server is used. For development and testing use write your node server code in the vite plugin inside `vite.config.ts` and copy this code into `web-ui/src/server/server.js` for production (alternatively create a new file that exports a function and call it in both then develop from that file).
```

            type: "button",
            action: () => {
              // action is executed on button press
              console.log("Button clicked");
              return "Success"; // string return value is printed in a toast
            },
          },
        },
        Advanced: {
          booleanSwitch: {
            type: "switch",
            description: "This is a switch", // fields with descriptions get a question mark next to their label
          },
          selectOptions: {
            type: "select",
            options: [
              { value: "1", label: "Option 1" },
              { value: "2", label: "Option 2" },
              { value: "3", label: "Option 3" },
            ],
          },
        },
      },
    });

    console.log(settings.groups.Advanced.booleanSwitch.value); // logs the state of the switch
    ```

- The second script tag is where your code should go as it is run client side when the component is rendered. The settings can be accessed here without an import and to listen for changes wrap the value you want to check in either a `$derived` or `$effect` rune. Furthermore, use `toast('message')` to send a toast from your widget.

> - The remainder of the file is where mark down goes. Tailwind classes can be used on widgets for styling or a `<style></style>` block.

Reviewer: ```suggestion
- The remainder of the file is where mark down goes. Tailwind classes can be used on widgets for styling or a `<style></style>` block, although the former is strongly preferred if possible.
```
